saltswap/record.py

```diff
@@ -0,0 +1,176 @@
+
+        # Scalar quantities with their units
+        for variable in variable_states:
+            var = sample_state_group.createVariable(variable, 'f8', ('iteration', 'scalar'))
+            var.unit = variable_states[variable].get_name()
+
+        # Which molecule is which
+        sample_state_group.createDimension('identity', len(swapper.mutable_residues))
+        sample_state_group.createVariable('identities', 'i8', ('iteration', 'identity'))
+
+        # How many water, cations, and anions there are.
+        sample_state_group.createDimension('species', len(swapper.get_identity_counts()))
+        sample_state_group.createVariable('species counts', 'i8', ('iteration', 'species'))
+
+        # The cumulative work for each NCMC step in thermal units (i.e. unitless)
+        sample_state_group.createDimension('attempt', None)
+        sample_state_group.createDimension('npert', swapper.npert + 1)
+        var = sample_state_group.createVariable('cumulative work', 'f8', ('iteration', 'attempt', 'npert'))
+        var.unit = 'unitless'
+        # The corresponding proposal for the NCMC protocol work
+        sample_state_group.createDimension('proposal', 2)
+        sample_state_group.createVariable('proposal', 'i8', ('iteration', 'attempt', 'proposal',))
+
+        # The number of saltswap moves that have been accepted
+        sample_state_group.createVariable('naccepted', 'i8', ('iteration', 'attempt', 'scalar',))
+
+        self.ncfile.sync()
+
+    def create_netcdf(self, swapper, variable_dic=None):
+        """
+        Create a
+        Parameters
+        ----------
+        swapper: saltswap.swapper
+            the driver for saltswap
+        variable_dic: dic
+            dictionary containing additional parameters to be stored
+        """
+        self.init_control_variables(swapper, variable_dic)
+        self.init_sample_state_variables(swapper)
+
+        self.ncfile.sync()
+
+        return self.ncfile
+
+def record_netcdf(ncfile, context, swapper, iteration, attempt=0, sync=True):
+    """
+    Store the variables in the context and swapper objects.
+
+    Parameters
+    ----------
+    ncfile: netCDF4._netCDF4.Dataset
+        Netcdf data object that has already been initialized to record saltswap simulation data.
+    context: simtk.openmm.openmm.Context
+        Contains the OpenMM simulation data and parameters
+    swapper: saltswap.swapper
+        the driver for saltswap
+    """
+    # Openmm state information
+    state = context.getState(getPositions=True, getEnergy=True, enforcePeriodicBox=True)
+
+    volume = state.getPeriodicBoxVolume()
+    var_units = variable_states['volume']
+    ncfile.groups['Sample state data']['volume'][iteration, :] = volume.value_in_unit(var_units)
+
+    potential_energy = state.getPotentialEnergy()
+    var_units = variable_states['potential energy']
+    ncfile.groups['Sample state data']['potential energy'][iteration, :] = potential_energy .value_in_unit(var_units)
```

Comment:
Is there a reason you need to use a slice `:` here? What is the second dimension of that array?